### PR TITLE
Twl/the one with kv storage query ranges internal writes kv del

### PR DESCRIPTION
**Pull request by Tadeuchi**

### CODE REVIEW

File: src/contract/states/ContractInteractionState.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    sortKey?: string,
+    options?: SortKeyCacheRangeOptions
+  ): Promise<Map<string, unknown>> {
+    const storage = this._warp.kvStorageFactory(contractTxId);
+    return storage.kvMap(sortKey, options);
+  }
+
+  async commit(interaction: GQLNodeInterface): Promise<void> {
+    if (interaction.dry) {
+      return await this.rollbackKVs();
```
Comment (ppedziwiatr) [inline src/contract/states/ContractInteractionState.ts:52]: add `this.reset()` here?

**Review by ppedziwiatr — CHANGES_REQUESTED** [review]

File: src/cache/impl/LevelDbCache.ts [hunk 2]
```diff
@@ -142,10 +204,10 @@ export class LevelDbCache<V> implements SortKeyCache<V> {
     await this.db.open();
     const keys = await this.db.keys().all();
 
-    for (const key of keys) {
-      // default key format used by sub-levels:
-      // !<contract_tx_id (43 chars)>!<sort_key>
-      const sortKey = key.substring(45);
+    for (const joinedKey of keys) {
+      // default joined key format used by sub-levels:
+      // <separator><contract_tx_id (43 chars)><separator><sort_key>
```
[the remark this answers is not in the document]
Comment (Tadeuchi) [reply src/cache/impl/LevelDbCache.ts:250]: changed to `const sortKey = joinedKey.split(this.subLevelSeparator)[1];`

File: src/cache/SortKeyCacheRangeOptions.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+export interface SortKeyCacheRangeOptions {
+  gte?: string;
```
[the remark this answers is not in the document]
Comment (Tadeuchi) [reply src/cache/SortKeyCacheRangeOptions.ts:2]: yes, correct

File: src/contract/states/ContractInteractionState.ts (hunk 1 above)
Comment (Tadeuchi) [reply src/contract/states/ContractInteractionState.ts]: as you wish

File: src/cache/impl/LevelDbCache.ts [hunk 4]
```diff
@@ -129,6 +175,22 @@ export class LevelDbCache<V> implements SortKeyCache<V> {
     }
   }
 
+  begin() {
+    this._rollbackBatch = this.db.batch();
+  }
+
+  async rollback() {
+    if (this._rollbackBatch && this._rollbackBatch.length > 0) {
```
[the remark this answers is not in the document]
Comment (Tadeuchi) [reply src/cache/impl/LevelDbCache.ts]: added some simple checking that the previous transaction has finished
there is some marker saved on kv.begin 
and deleted on commit or rollback
on open the kvStorage should throw an error if transaction marker found

it seems that the tx id is not needed for this

**Review by ppedziwiatr — APPROVED** [review]

**Pull request merged**